XGBoost sparse matrix decision boundary rule ehnanced with categorica…

Pscheidl opened the pull request:
…l column counter.

Not an actual PR now. More like a playground.

wendycwong commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoost.java:
```diff
@@ -268,17 +271,28 @@ final void buildModelImpl() {
+        denominator = denominator.multiply(BigDecimal.valueOf(_train.numRows()), MathContext.DECIMAL128);
+        double fillRatio = nonZeroCount.divide(denominator, MathContext.DECIMAL128).doubleValue();
         Log.info("fill ratio: " + fillRatio);
         model._output._sparse = fillRatio < 0.5 || ((_train.numRows() * (long) _train.numCols()) > Integer.MAX_VALUE);
```
Pavel:  Don't quite understand the logic here.  Okay, basically, we want to switch over to a sparse representation (lists denoting the row and column indices of an element) instead of using a double array.  Hence, we have this fillRatio thing to guide us (may want to change this ratio depending on the memory requirement of both representations).  However, what is the second OR condition for?

wendycwong commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoost.java:
```diff
@@ -268,17 +271,28 @@ final void buildModelImpl() {
-        fillRatio /= col;
+        BigDecimal denominator = BigDecimal.valueOf(numOneHotEncodedCols).add(BigDecimal.valueOf(nonCategoricalCols));
+        denominator = denominator.multiply(BigDecimal.valueOf(_train.numRows()), MathContext.DECIMAL128);
+        double fillRatio = nonZeroCount.divide(denominator, MathContext.DECIMAL128).doubleValue();
```
The calculations look good to me.  Thanks.

wendycwong commented:
Pavel:

Would like to propose to change the fill_ratio to be 1/3 or less.  Here is my calculation:

let m by number of rows and n be number of columns (expanded for enums).  The fill-ratio is num_nz/(m*n).  

The dense storage for a dataset is a double[][] array, this means the memory allocated is m*n*memory used to store one cell (32 or 64 bits).  

The storage for sparse matrix is 3 double arrays, one for column index, one for row index and one for the cell value.  Hence, the ratio of memory use between these two methods are:

ratio = 3*num_nz*memory to store one cell /(m*n * memory to store one cell) = 3*frac.

We see a memory saving using sparse representation if ratio < 1 and this implies 3*frac < 1 and hence frac < 1/3.

Let me know what you think.   

Wendy

Pscheidl commented:
@wendycwong 

The data structures are indeed as you say.

```java
  /**
   * Create DMatrix from Sparse matrix in CSR/CSC format.
   * 2D arrays since underlying array can accomodate that many elements.
   * @param headers The row index of the matrix.
   * @param indices The indices of presenting entries.
   * @param data The data content.
   * @param st  Type of sparsity.
   * @param shapeParam   when st is CSR, it specifies the column number, otherwise it is taken as
   *                     row number
   * @param shapeParam2  when st is CSR, it specifies the row number, otherwise it is taken as
   *                     column number
   * @param ndata   number of nonzero elements
   * @throws XGBoostError
   */
  public DMatrix(long[][] headers, int[][] indices, float[][] data, SparseType st,
                 int shapeParam, int shapeParam2, long ndata)
```

Your optimization is correct and I agree.

michalkurka commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoost.java:
```diff
@@ -268,19 +272,33 @@ final void buildModelImpl() {
       } else if (_parms._dmatrix_type == XGBoostModel.XGBoostParameters.DMatrixType.dense) {
         model._output._sparse = false;
       } else {
-        float fillRatio = 0;
```
can you please put this in a method?

Pscheidl replied:
:ok:  Done

Pscheidl commented:
@wendycwong No failures ? The matrix should be treated as sparse in many cases. 

My tests show improvements on airlines data (both small and full) - matrices are now considered to be sparse corectly. Maybe I **should write a test** for it. We have that information.

wendycwong commented:
Pavel:  Writing a test for it will be good.

wendycwong commented on h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java:
```diff
@@ -605,6 +605,68 @@ public void ProstateRegression() {
+
+      // GLEASON used as predictor variable, numeric variable, dense
+      model = new hex.tree.xgboost.XGBoost(parms).trainModel().get();
+      assertFalse(model._output._sparse);
```
How about adding one test to detect sparsity?  Randomly generate enum columns only with high cardinality.  You can use this function: TestUtil.generate_enum_only?

Pscheidl commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoost.java:
```diff
@@ -345,6 +337,39 @@ final void buildModelImpl() {
+      Log.info("fill ratio: " + fillRatio);
+
+      return fillRatio < FILL_RATIO_THRESHOLD
+          || ((_train.numRows() * totalColumns) > Integer.MAX_VALUE);
```
Honestly, I do not understand it's purpose too. This is original line by Mateusz and I did not want this PR to affect original parts we could rely on.

Pscheidl commented:
1. Rebased everything on top of latest master (there was a trivial conflict in `XGBoostTest` required to be resolved)
2. Added sparsity/density detection test like @wendycwong suggested.
